src/rtc_stream.cpp

```diff
@@ -0,0 +1,418 @@
+		auto self = weak_this.lock();
+		if (!self) return;
+
+		post(m_io_context, std::bind(&rtc_stream_impl::on_buffered_low
+			, self
+			, error_code{}
+		));
+	});
+
+	m_data_channel->onClosed([this, weak_this]()
+	{
+		// Warning: this is called from another thread
+		auto self = weak_this.lock();
+		if (!self) return;
+
+		post(m_io_context, std::bind(&rtc_stream_impl::cancel_handlers
+			, self
+			, boost::asio::error::connection_reset
+		));
+	});
+}
+
+void rtc_stream_impl::close()
+{
+	if (m_data_channel && !m_data_channel->isClosed())
+		m_data_channel->close();
+
+	cancel_handlers(boost::asio::error::operation_aborted);
+}
+
+void rtc_stream_impl::on_available(error_code const& ec)
+{
+	if (!m_read_handler) return;
+
+	if (ec)
+	{
+		clear_read_buffers();
+		post(m_io_context, std::bind(std::exchange(m_read_handler, nullptr), ec, 0));
+		return;
+	}
+
+	// Fulfil pending read
+	issue_read();
+}
+
+void rtc_stream_impl::on_buffered_low(error_code const& ec)
+{
+	if (!m_write_handler) return;
+
+	std::size_t const bytes_written = ec ? 0 : m_write_buffer_size;
+
+	m_write_buffer.clear();
+	m_write_buffer_size = 0;
+    post(m_io_context, std::bind(std::exchange(m_write_handler, nullptr), ec, bytes_written));
+}
+
+bool rtc_stream_impl::is_open() const
+{
+	return m_data_channel && m_data_channel->isOpen();
+}
+
+std::size_t rtc_stream_impl::available() const
+{
+	return m_incoming.size() + (m_data_channel ? m_data_channel->availableAmount() : 0);
+}
+
+rtc_stream::endpoint_type rtc_stream_impl::remote_endpoint(error_code& ec) const
+{
+    if (!is_open())
+    {
+        ec = boost::asio::error::not_connected;
+        return {};
+    }
+
+	auto addr = m_peer_connection->remoteAddress();
+	if (!addr)
+	{
+		ec = boost::asio::error::operation_not_supported;
+		return {};
+	}
+
+	// The format is always "address:port" without brackets around the address,
+	// therefore splitting on the last ':' works to separate address and port.
+	std::size_t pos = addr->find_last_of(':');
+	if (pos == std::string::npos)
+	{
+		ec = errors::parse_failed;
+		return {};
+	}
+
+	return endpoint_type(ip::make_address(addr->substr(0, pos), ec)
+			, std::uint16_t(std::stoul(addr->substr(pos+1))));
+}
+
+rtc_stream::endpoint_type rtc_stream_impl::local_endpoint(error_code& ec) const
+{
+	if (!is_open())
+    {
+        ec = boost::asio::error::not_connected;
+        return {};
+    }
+
+	auto addr = m_peer_connection->localAddress();
+	if (!addr)
+	{
+		ec = boost::asio::error::operation_not_supported;
+		return {};
+	}
+
+	// The format is always "address:port" without brackets around the address,
+	// therefore splitting on the last ':' works to separate address and port.
+	std::size_t pos = addr->find_last_of(':');
+	if (pos == std::string::npos)
+	{
+		ec = errors::parse_failed;
+		return {};
+	}
+
+	return endpoint_type(ip::make_address(addr->substr(0, pos), ec)
+			, std::uint16_t(std::stoul(addr->substr(pos+1))));
```

Comment:
`std::stoul()` will throw if it fails to convert. The caller of this function (`local_endpoint()`) might expect `ec` to be set instead